terraform/provider/resource_teleport_role.go

```diff
@@ -1,61 +1,145 @@
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	// roleSpec := services.RoleSpecV3{}
+	name, err := getRoleName(d)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	// role, err := services.NewRole(name, roleSpec)
-	// if err != nil {
-	// 	return trace.Wrap(err)
-	// }
+	role, err := types.NewRole(name, types.RoleSpecV3{})
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	// FIXME: Client.CreateRole() is not implemented yet in api/client
-	// err = client.CreateRole(role)
-	// if err != nil {
-	// 	return trace.Wrap(err)
-	// }
+	roleV3, ok := role.(*types.RoleV3)
+	if !ok {
+		return diag.FromErr(trace.Errorf("failed to convert created role to types.RoleV3"))
+	}
 
-	d.SetId(name)
-	return nil
+	err = tfschema.UnmarshalRoleV3(d, roleV3)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	err = c.UpsertRole(ctx, roleV3)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	d.SetId(roleV3.GetName())
+
+	return resourceRoleRead(ctx, d, m)
 }
 
-func resourceTeleportRoleRead(d *schema.ResourceData, m interface{}) error {
-	// client := m.(*auth.Client)
-	// name := d.Get("name").(string)
+// resourceRoleRead reads Teleport role
+func resourceRoleRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	return nil
+	roleID := d.Id()
+
+	_, err = c.GetRole(ctx, roleID)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	return diag.Diagnostics{}
 }
 
-func resourceTeleportRoleDelete(d *schema.ResourceData, m interface{}) error {
-	// client := m.(*auth.Client)
-	// name := d.Get("name").(string)
+// resourceRoleUpdate updates Teleport role
+func resourceRoleUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	return nil
+	roleID := d.Id()
+
+	role, err := c.GetRole(ctx, roleID)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	roleV3, ok := role.(*types.RoleV3)
+	if !ok {
+		return diag.FromErr(trace.Errorf("failed to convert created role to types.RoleV3"))
+	}
+
+	err = tfschema.UnmarshalRoleV3(d, roleV3)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	err = c.UpsertRole(ctx, roleV3)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	return resourceRoleRead(ctx, d, m)
 }
 
-func resourceTeleportRoleExists(d *schema.ResourceData, m interface{}) (bool, error) {
-	// client := m.(*auth.Client)
-	// name := d.Get("name").(string)
+// resourceRoleDelete deletes Teleport role
+func resourceRoleDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	roleID := d.Id()
+	err = c.DeleteRole(ctx, roleID)
```

Comment:
Should this check for role existence first?

Reply:
I think, no. Terraform has it's own state which is considered as the primary source of truth. So, if there would be a difference between Terraform and real state, we need to consider it as error.